README.md
<table>
<tbody>
<tr class="odd">
<td># author: Silvana M. Pesenti</td>
</tr>
<tr class="even">
<td># date: June 01, 2019</td>
</tr>
<tr class="odd">
<td>output:</td>
</tr>
<tr class="even">
<td>rmarkdown::html_vignette:</td>
</tr>
<tr class="odd">
<td>md_extensions: [</td>
</tr>
<tr class="even">
<td>“-autolink_bare_uris”</td>
</tr>
<tr class="odd">
<td>]</td>
</tr>
</tbody>
</table>

<!-- README.md is generated from README.Rmd. Please edit that file -->

# SWIM - A Package for Sensitivity Analysis

[![Travis-CI Build
Status](https://travis-ci.org/spesenti/SWIM.svg?branch=master)](https://travis-ci.org/spesenti/SWIM)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/SWIM)](https://cran.r-project.org/package=SWIM)
[![downloads](https://cranlogs.r-pkg.org/badges/grand-total/SWIM)](https://cran.r-project.org/package=SWIM)

The SWIM package provides weights on simulated scenarios from a
stochastic model, such that stressed model components (random variables)
fulfil given probabilistic constraints (e.g. specified values for risk
measures), under the new scenario weights. Scenario weights are selected
by constrained minimisation of the relative entropy to the baseline
model. The SWIM package is based on the papers Pesenti S.M, Millossovich
P., Tsanakas A. (2019) [“Reverse Sensitivity Testing: What does it take
to break the model”](https://openaccess.city.ac.uk/id/eprint/18896/) and
and Pesenti S.M. (2021) [“Reverse Sensitivity Analysis for Risk
Modelling”](https://www.ssrn.com/abstract=3878879).

## Vignette

The Vignette of the SWIM package is available in html format
utstat.toronto.edu/pesenti/SWIMVignette/ and as pdf
<https://openaccess.city.ac.uk/id/eprint/25845/>.

## Installation

The SWIM package can be installed from
[CRAN](https://CRAN.R-project.org/package=SWIM) :

> <https://CRAN.R-project.org/package=SWIM>;

alternatively from [GitHub](https://github.com/spesenti/SWIM):

> <https://github.com/spesenti/SWIM>

## Scope of the SWIM package

The SWIM package provides sensitivity analysis tools for stressing model
components (random variables). Implemented stresses using the relative
Entropy (Kullback-Leibler divergence) are:

<table>
<thead>
<tr class="header">
<th>R functions</th>
<th>Stress</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>stress()</code></td>
<td>A wrapper for the <code>stress_</code> functions</td>
</tr>
<tr class="even">
<td><code>stress_VaR()</code></td>
<td>VaR risk measure, a quantile</td>
</tr>
<tr class="odd">
<td><code>stress_VaR()</code></td>
<td>VaR risk measure, a quantile</td>
</tr>
<tr class="even">
<td><code>stress_VaR_ES()</code></td>
<td>VaR and ES risk measures</td>
</tr>
<tr class="odd">
<td><code>stress_mean()</code></td>
<td>means</td>
</tr>
<tr class="even">
<td><code>stress_mean_sd()</code></td>
<td>means and standard deviations</td>
</tr>
<tr class="odd">
<td><code>stress_moment()</code></td>
<td>moments, functions of moments</td>
</tr>
<tr class="even">
<td><code>stress_prob()</code></td>
<td>probabilities of intervals</td>
</tr>
<tr class="odd">
<td><code>stress_user()</code></td>
<td>user defined scenario weights</td>
</tr>
</tbody>
</table>

Implemented stresses using the 2-Wasserstein distance are:

<table>
<thead>
<tr class="header">
<th>R functions</th>
<th>Stress</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>stress_wass()</code></td>
<td>A wrapper for the <code>stress_w</code> functions</td>
</tr>
<tr class="even">
<td><code>stress_RM_w()</code></td>
<td>Risk measure</td>
</tr>
<tr class="odd">
<td><code>stress_RM_mean_sd_w()</code></td>
<td>Risk measure, means and standard deviations</td>
</tr>
<tr class="even">
<td><code>stress_HARA_RM_w()</code></td>
<td>Risk measure and HARA utility</td>
</tr>
<tr class="odd">
<td><code>stress_mean_sd_w()</code></td>
<td>means and standard deviations</td>
</tr>
<tr class="even">
<td><code>stress_mean_w()</code></td>
<td>means</td>
</tr>
</tbody>
</table>

Implemented functions allow to graphically display the change in the
probability distributions under different stresses and the baseline
model as well as calculating sensitivity measures.

## Example - Stressing the VaR of a portfolio

Consider a portfolio Y = X1 + X2 + X3 + X4 + X5, where (X1, X2, X3, X4,
X5) are correlated normally distributed with equal mean and different
standard deviations. We stress the VaR (quantile) of the portfolio loss
Y at levels 0.75 and 0.9 with an increase of 10%.

     # simulating the portfolio 
    library(SWIM)
    set.seed(0)
    SD <- c(70, 45, 50, 60, 75)
    Corr <- matrix(rep(0.5, 5^2), nrow = 5) + diag(rep(1 - 0.5, 5))
    x <- mvtnorm::rmvnorm(10^5, 
       mean =  rep(100, 5), 
       sigma = (SD %*% t(SD)) * Corr)
    data <- data.frame(rowSums(x), x)
    names(data) <- c("Y", "X1", "X2", "X3", "X4", "X5")
     # stressing the portfolio 
    rev.stress <- stress(type = "VaR", x = data, 
       alpha = c(0.75, 0.9), q_ratio = 1.1, k = 1)
    #> Stressed VaR specified was 722.9387 , stressed VaR achieved is 722.9378
    #> Stressed VaR specified was 878.859 , stressed VaR achieved is 878.8296

Summary statistics of the baseline and the stressed model can be
obtained via the `summary()` method.

    #> $base
    #> 
    #> 
    #> |            |      Y|     X1|     X2|     X3|     X4|     X5|
    #> |:-----------|------:|------:|------:|------:|------:|------:|
    #> |mean        | 500.18| 100.14| 100.01|  99.93|  99.98| 100.13|
    #> |sd          | 232.13|  69.79|  44.93|  49.83|  59.85|  74.76|
    #> |skewness    |   0.00|  -0.01|   0.00|   0.01|  -0.01|   0.01|
    #> |ex kurtosis |  -0.04|  -0.03|  -0.02|  -0.01|  -0.02|   0.00|
    #> |1st Qu.     | 342.56|  53.24|  69.70|  66.23|  59.45|  49.72|
    #> |Median      | 500.45| 100.16| 100.06| 100.11| 100.23|  99.94|
    #> |3rd Qu.     | 657.22| 147.33| 130.31| 133.51| 140.33| 150.80|
    #> 
    #> $`stress 1`
    #> 
    #> 
    #> |            |      Y|     X1|     X2|     X3|     X4|     X5|
    #> |:-----------|------:|------:|------:|------:|------:|------:|
    #> |mean        | 534.03| 108.20| 104.85| 105.38| 106.70| 108.89|
    #> |sd          | 245.47|  72.32|  46.35|  51.48|  61.90|  77.60|
    #> |skewness    |  -0.06|  -0.04|  -0.03|  -0.02|  -0.04|  -0.02|
    #> |ex kurtosis |  -0.29|  -0.13|  -0.10|  -0.10|  -0.12|  -0.11|
    #> |1st Qu.     | 361.73|  58.81|  73.38|  70.21|  64.41|  55.97|
    #> |Median      | 532.15| 108.59| 105.08| 105.65| 107.33| 109.02|
    #> |3rd Qu.     | 722.94| 158.24| 136.71| 140.66| 149.34| 162.58|
    #> 
    #> $`stress 2`
    #> 
    #> 
    #> |            |      Y|     X1|     X2|     X3|     X4|     X5|
    #> |:-----------|------:|------:|------:|------:|------:|------:|
    #> |mean        | 524.20| 105.87| 103.44| 103.82| 104.75| 106.33|
    #> |sd          | 249.61|  73.14|  46.81|  52.01|  62.57|  78.46|
    #> |skewness    |   0.09|   0.04|   0.04|   0.05|   0.04|   0.06|
    #> |ex kurtosis |  -0.19|  -0.09|  -0.06|  -0.06|  -0.08|  -0.07|
    #> |1st Qu.     | 352.14|  55.99|  71.60|  68.34|  62.00|  52.93|
    #> |Median      | 516.03| 104.72| 102.95| 103.26| 104.12| 104.94|
    #> |3rd Qu.     | 687.58| 155.33| 134.95| 138.82| 146.79| 159.24|

Visual display of the change of empirical distribution functions of the
portfolio loss Y from the baseline to the two stressed models.

    library(spatstat)
    plot_cdf(object = rev.stress, xCol = 1, base = TRUE)

<img src="man/figures/README-plot-cdf-1.png" width="100%" />

### Sensitivity and importance rank of portfolio components

Sensitivity measures allow to assess the importance of the input
components. Implemented sensitivity measures are the Kolmogorov
distance, the Wasserstein distance and *Gamma*. *Gamma*, the *Reverse
Sensitivity Measure*, defined for model component Xi, i = 1, …, 5, and
scenario weights w by

*Gamma* = ( E(Xi \* w) - E(Xi) ) / c,

where c is a normalisation constant such that |*Gamma*| &lt;= 1, see
<https://doi.org/10.1016/j.ejor.2018.10.003>. Loosely speaking, the
Reverse Sensitivity Measure is the normalised difference between the
first moment of the stressed and the baseline distributions of Xi.

    knitr::kable(sensitivity(rev.stress, type = "all"), digits = 2)
    #> Warning in sensitivity(rev.stress, type = "all"): No s passed in. Using Gamma sensitivity instead.

<table>
<thead>
<tr class="header">
<th style="text-align: left;">stress</th>
<th style="text-align: left;">type</th>
<th style="text-align: right;">Y</th>
<th style="text-align: right;">X1</th>
<th style="text-align: right;">X2</th>
<th style="text-align: right;">X3</th>
<th style="text-align: right;">X4</th>
<th style="text-align: right;">X5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">stress 1</td>
<td style="text-align: left;">Gamma</td>
<td style="text-align: right;">1.00</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.74</td>
<td style="text-align: right;">0.75</td>
<td style="text-align: right;">0.77</td>
<td style="text-align: right;">0.81</td>
</tr>
<tr class="even">
<td style="text-align: left;">stress 2</td>
<td style="text-align: left;">Gamma</td>
<td style="text-align: right;">1.00</td>
<td style="text-align: right;">0.79</td>
<td style="text-align: right;">0.74</td>
<td style="text-align: right;">0.75</td>
<td style="text-align: right;">0.77</td>

NEWS.md

# SWIM 1.0.0 - current version on GitHub

## Major changes: Additional functions and features

 - Wasserstein distance
 
     - `stress_wass()`:
        - A wrapper for the stress functions using the 2-Wasserstein distance
     - `stress_RM_w()`:
        - a stressed model component (random variable) fulfills a 
        constraint on its risk measure defined by a gamma function.
     - `stress_RM_mean_sd_w()`:
        - a stressed model component (random variable) fulfills a 
        constraint on its mean, standard deviation, and risk measure 
        defined by a gamma function.
     - `stress_HARA_RM_w()`:
        - a stressed model component (random variable) fulfills a
        constraint on its HARA utility defined by a, b and eta parameter
        and risk measure defined by a gamma function.
    - `stress_mean_sd_w()`:
        - a stressed model component (random variable) fulfills a
        constraint on its mean and standard deviation.
    - `stress_mean_w()`: 
        - a stressed model component (random variable) fulfills a
        constraint on its mean.
      

 - Functions
   
     - `mean_stressed()`:
        - sample mean of chosen stressed model components, subject to the calculated scenario weights.
     - `sd_stressed()`:
        - sample standard deviation of chosen stressed model components, subject to the calculated scenario weights.
     - `var_stressed()`:
        - sample variance of chosen stressed model components, subject to the calculated scenario weights.
     - `cor_stressed()`:
        - sample correlation coefficient of chosen stressed model components, subject to the calculated scenario weights.
    - `cdf_stressed()`:
        - the empirical distribution function of a stressed model component (random variable) under the scenario weights. 
    - `rename_SWIM()`: 
        - Get a new SWIM object with desired names.
    
    
    
- Features

    - `stress()`:
        - A parameter "names" to all stress functions, which allows to name a stress differently than just "stress 1", "stress 2", etc.
        - A parameter "log" that allows users to inspect weights' statistics, including minimum, maximum, standard deviation, Gini coefficient, and entropy.
    - `sensitivity()`:
        - A parameter "p"  can be specified for the degree of Wasserstein distance.

## Minor changes

 - fix minor bug in `summary()`.
 - add `base` argument for `quantile_stressed()` and an error message if the input has `wCol` has dimension larger than 1.


# SWIM 0.2.2 - current version on CRAN

## Major changes: Additional functions and features

 - `plot_quantile()`:
    * the function plots the empirical quantile of model components, subject to 
      scenario weights.

 - `plot_weights()`:
    * the function plots the scenario weights of a stressed model against 
      model components.

 - `stress_moment()`:
    * add parameter "normalise" that allows to linearly normalise the values
    called by `nleqslv`.
    * the function prints a table with the required and achieved moments and the absolute and relative error.

 - `stress_VaR_ES()`:
    * add parameter "normalise" that allows to linearly normalise the values 
    before `uniroot` is applied.


## Minor changes

 - fix bug in merging different stress objects.


# SWIM 0.2.1 

## Minor changes

- add vignette
- fix bug in `merge()`. 
- fix bug in `sensitivity()`. 


# SWIM 0.2.0

## Major changes

### Additional functions and data sets

 - `VaR_stressed()`:
    * the function calculates the VaR of model components, subject to 
      scenario weights.

 - `ES_stressed()`:
    * the function calculates the ES of model components, subject to 
      scenario weights.
    
 - `credit_data`:
    * a data set containing aggregate losses from a credit portfolio,
      generated through a binomial credit model.

### Amendments to functions

 - `stress_VaR()`:
    * amendment to the calculation of scenario weights when the specified VaR cannot be achieved.
    * returns a message if the achieved VaR is not equal to the stressed VaR specified.
    * specs of the `SWIM` object contains the achieved VaR 
    * allowing for stressing VaR downwards
    
 - `stress_VaR_ES()`:
    * amendment analogous to the `stress_VaR()`.
    * returns a message if the achieved VaR is not equal to the stressed VaR specified.
    * specs of the `SWIM` object contains the achieved VaR 
    * allowing for stressing VaR and ES downwards

## Minor changes

 - `stress()`:   
    * parameter `x` can have missing column names.

 - `stress_moment()`:
    * additional parameter `show`; if `TRUE` (default is `FALSE`), the result of `nleqslv()` is printed.


R/class.R
 # Defines the class "SWIM"
  SWIM <- function(x = "x", new_weights = "new_weights", 
                       type = "type", specs = "specs"){
   mymodel <- list(
      x = x, 
      # vector, matrix or dataframe containing the underlying data
      new_weights = new_weights, 
      # list of eithter functions, that applied 
      # to the kth column of x providing the scenario weights; OR a 
      # vector containting the new_weights
      type  = type,
      # a list of characters each corresponding to a stress
      # one of ("VaR", "VaR ES", "prob", "moment", "mean", "mean sd", "user")
      specs = specs
      # a list with elements called "stress i".
      # all input varaibles of the stress and constraints according
      # to the stress. For example a stress on 
      # the VaR contains: k, alpha, q
      # names = names
   )   
   
   ## Name of the class
   attr(mymodel, "class") <- "SWIM"
   return(mymodel)
  }

  
  is.SWIM <- function(object) inherits(object, "SWIM")

 # Defines the class "SWIMw"
  SWIMw <- function(x = "x", h = "h", u="u", lam = "lam", new_weights = "new_weights", 
                    str_fY = 'str_fY', str_FY = 'str_FY', str_FY_inv = 'str_FY_inv',
                    gamma = 'gamma', type = "type", specs = "specs"){
     mymodel <- list(
        x = x, # vector, matrix or dataframe containing the underlying data
        h = h, # function, that applied to x, provide the bandwidths for KDE estimation
        u = u, # vector containing the gridspace on [0,1]
        lam = lam, # optimized lambda value(s)
        str_fY = str_fY, # function, that applied to the stressed column, provides the density
        str_FY = str_FY, # function, when applied to the stressed column, provides the cdf
        str_FY_inv = str_FY_inv, # function, when applied to the stressed column, provides the quantile
        gamma = gamma, # function that provides gamma used to calculate the risk measure (if applicable);
        new_weights = new_weights, # list of either functions, that applied 
        # to the k-th column of x providing the scenario weights; OR a 
        # vector containing the new_weights
        type  = type, # a list of characters each corresponding to a stress
        # one of ("RM", "mean sd", "RM mean sd", "HARA RM")
        specs = specs # a list with elements called "stress i".
        #
        # all input variables of the stress and constraints according
        # to the stress. For example a stress on 
        # the RM contains: k, alpha, q
     )   
     ## Name of the class
     attr(mymodel, "class") <- "SWIMw"
     return(mymodel)
  }
  
  is.SWIMw <- function(object) inherits(object, "SWIMw")
  
 #' Extracting from a Stressed Model
 #'
 #' Extracting the data (realisations of the stochastic model), the 
 #'     scenario weights, the functions generating the scenario weights, 
 #'     or the specifications of the stress from an object of class 
 #'     \code{SWIM} or \code{SWIMw}. 
 #' 
 #' @param object    A \code{SWIM} or \code{SWIMw} object.
 #' @inheritParams   summary.SWIM
 #'  
 #' @return \code{get_data}: A data.frame containing the realisations of 
 #'         the stochastic model on which the \code{object} is based.    
 #'
 #' @seealso \code{\link{SWIM}}
 #'                  
 #' @author Silvana M. Pesenti 
 #' @describeIn get_data extracting data.
 #' @export

  get_data <- function(object, xCol = "all"){
   if (!is.SWIM(object) && !is.SWIMw(object)) stop("Object not of class SWIM or SWIMw")
   if (length(xCol) ==1 && xCol == "all") {
      xdata <- as.matrix(object$x[, 1:ncol(object$x)])
      if(is.character(colnames(object$x)) | is.null(colnames(object$x))) colnames(xdata) <- colnames(object$x)
   } else{ 
#   if (length(xCol) ==1 && xCol == "all" && is.null(colnames(object$x))) {xCol <- 1:ncol(object$x) 
#      } else if (length(xCol) ==1 && xCol == "all") {xCol <- colnames(object$x)} 
   if (is.null(xCol)) stop("invalid 'xCol' argument")
   if (is.numeric(xCol) && !(any(xCol %in% 1:ncol(object$x)))) stop("invalid 'xCol' argument")
   if (is.character(xCol) && !(any(xCol %in% colnames(object$x)))) stop("invalid 'xCol' argument")
   xdata <- as.matrix(object$x[, xCol])
   if (is.character(xCol[xCol])) colnames(xdata) <- xCol
   if (is.numeric(xCol) && is.null(colnames(object$x)[xCol])) {
      colnames(xdata) <- paste("X", as.character(xCol), sep = "")
   } else if (is.numeric(xCol) && is.character(colnames(object$x)[xCol])) {
      colnames(xdata) <- colnames(object$x)[xCol]
   }}
   return(xdata)
  }

 #' @describeIn get_data extracting scenario weights. 
 #'
 #' @return \code{get_weights:} A data.frame containing the scenario 
 #'     weights of the \code{object}. Columns corresponds 
 #'     to different stresses.
 #'         
 #' @examples 
 #' ## continuing example in stress_VaR
 #' set.seed(0)
 #' x <- as.data.frame(cbind(
 #'      "normal" = rnorm(1000), 
 #'      "gamma" = rgamma(1000, shape = 2)))
 #'   res1 <- stress(type = "VaR", x = x, 
 #'                  alpha = 0.9, q_ratio = 1.05, k = 1)
 #' 
 #' ## returning the underlying data
 #' all(get_data(res1) == x)
 #'  ## the scenario weights
 #' w <- get_weights(res1) 
 #' get_weightsfun(res1)
 #' get_specs(res1)
 #'   
 #' ## now add a stress on the means of both variables
 #' res1 <- stress(type = "mean", x = res1, k = 1:2, new_means = c(0.5,1.5))
 #' get_specs(res1)
 #' ## the required moments for a stress of type "mean" are not displayed 
 #' ## the type of stress and the specs for the second stress can be 
 #' ## extracted directly from the SWIM object.
 #' res1$type[[2]]
 #' res1$specs[[2]]
 #'                                              
 #'                                                               
 #'                                                                                                 
 #' @export

  get_weights <- function(object, wCol = "all"){
   if (!is.SWIM(object) && !is.SWIMw(object)) stop("Object not of class SWIM or SWIMw")
   x_data <- get_data(object)
   m <- length(object$type)
   new_weights <- matrix(0, nrow = nrow(x_data), ncol = m)
   for(i in 1:m){
    if (!is.function(object$new_weights[[i]])) {
     new_weights[, i] <- object$new_weights[[i]]
    } else {
     k <- object$specs[[i]]$k
     new_weights[, i] <- object$new_weights[[i]](x_data[, k])
    }
   }
   colnames(new_weights) <- names(object$specs)
   if (is.character(wCol) && wCol == "all") return (new_weights)
   return(new_weights[, wCol])
  }

 #' @describeIn get_data extracting weight functions.
 #'
 #' @return \code{get_weightsfun}: A list containing functions, which, 
 #'     when applied to a column of the data, generate the 
 #'     scenario weights of the \code{object}. The corresponding stressed 
 #'     columns can be obtained via \code{get_specs}.\cr
 #'     
 #'     Use \code{\link{get_weights}} if the \code{SWIM} object only contains 
 #'     scenario weights and not a list of functions.
 #'         
 #' @export

  get_weightsfun <- function(object, wCol = "all"){
   if (!is.SWIM(object) && !is.SWIMw(object)) stop("Object not of class SWIM or SWIMw")
   if (!is.function(object$new_weights[[1]]))
      stop("New_weights is not a function, use get_weights() instead.") 
   if (is.character(wCol) && wCol == "all") return (object$new_weights)
   return(object$new_weights[wCol])
  }

 #' @describeIn get_data extracting information of the stress.
 #'
 #' @return \code{get_specs}: A data.frame containing specifications 
 #'         of the stresses with each row corresponding to a different 
 #'         stress. Only a selection of the specifications is returned; 
 #'         however, all input variables are stored in the \code{object}.
 #'         See also \code{\link{SWIM}}.
 #' @export

  get_specs <- function(object, wCol = "all"){
   if (!is.SWIM(object) && !is.SWIMw(object)) stop("Object not of class SWIM or SWIMw.")
   .type <- object$type 
   .specs <- data.frame()
   if (is.SWIM(object)){
     for (i in 1:length(.type)){  
       if (.type[i] %in% c("VaR", "VaR ES", "user", "prob")){
         if (.type[[i]] == "prob" && length(object$specs[[i]]$prob) > 1){
           .specs <- plyr::rbind.fill(.specs, as.data.frame(t(unlist(object$specs[[i]]))))       
         } else {
           .specs <- plyr::rbind.fill(.specs, as.data.frame(object$specs[[i]], stringsAsFactors = FALSE)) 
         }
       } else if (.type[i] %in% c("moment", "mean", "mean sd")){
         k <- paste(object$specs[[i]]$k, collapse = " ")
         .specs <- plyr::rbind.fill(.specs, as.data.frame(k))
       } else stop("Object contains wrong type.")
    }
  } else {
    # Wasserstein
    for (i in 1:length(.type)){
     if (.type[i] %in% c("ES", "ES mean sd", "HARA ES")) {
       .specs <- plyr::rbind.fill(.specs, as.data.frame(object$specs[[i]], stringsAsFactors = FALSE)[1:3]) # Only include k, alpha, and q
     } else if (.type[i] %in% c("RM", "RM mean sd", "HARA RM")) {
       .specs <- plyr::rbind.fill(.specs, as.data.frame(object$specs[[i]], stringsAsFactors = FALSE)[c(1,3)]) # Only include k and q
     } else if (.type[i] %in% c("mean sd", "mean")){
       k <- paste(object$specs[[i]]$k, collapse = " ")
       .specs <- plyr::rbind.fill(.specs, as.data.frame(k))
    } else stop("Object contains wrong type.")
    }
  }
   .type <- t(as.data.frame(.type))
   .specs <- cbind(.type, .specs) 
   colnames(.specs)[1] <- "type"
   rownames(.specs) <- names(object$specs)
   return(.specs)
  }

  #' Rename Stressed Models
  #'     
  #' @details Get a new \code{SWIM} object with desired \code{name}
  #' 
  #' @param object A \code{SWIM} or \code{SWIMw} object
  #' @param names   Character vector, the new names of k-th stressed model.
  #' @param k   Numeric vector, the k-th stressed model of object to rename. (\code{default = 1}).
  #'  
  #' @return An renamed object of class \code{SWIM} containing:
  #'   \itemize{
  #'     \item \code{x}, a data.frame containing the data;
  #'     \item \code{new_weights}, a list, each component corresponds to 
  #'    a different stress and is either a vector of scenario weights or a
  #'    function, that applied to a column of \code{x}, generates the 
  #'    vectors of scenario weights; 
  #'     \item \code{type}, a list, each component corresponds to a 
  #'    different stress and specifies the type of the stress;
  #'     \item \code{specs}, a list, each component corresponds to 
  #'    a different stress and contains a list with the specifications 
  #'    of what has been stressed.
  #'   }
  #' See \code{\link{SWIM}} for details.
  #' 
  #' @examples
  #' set.seed(0)
  #' x <- as.data.frame(cbind(
  #'   "normal" = rnorm(1000),
  #'   "gamma" = rgamma(1000, shape = 2)))
  #' res1 <- stress(type = "VaR", x = x,
  #'   alpha = 0.9, q_ratio = 1.05)
  #' res1 <- rename(res1, "VaR_09", 1)
  #' 
  #' @author Kent Wu 
  #'
  #' @export
  
  rename <- function(object, `names`, k=1){
    if (!is.SWIM(object) & !is.SWIMw(object)) stop("Object not of class SWIM or SWIMw")
    if (length(names) != length(k)) stop("The number of names should be equal to k")
    temp <- names(object$new_weights)
    if (length(temp) > length(k) | length(temp) > length(names)) stop("The number of names or k exceeds the number of stressed models")
    temp[k] <- names
    names(object$new_weights) <- temp
    names(object$specs) <- temp
    object <- object
    return(object)
  }
  
 #' Merging Two Stressed Models
 #'
 #' This function is a \code{method} for an object of class 
 #'     \code{SWIM}.
 #'     
 #' @details Merges two objects of class \code{SWIM}, that 
 #'     are based on the same data. \cr 
 #' 
 #' @param x,y    Objects of class \code{SWIM}.
 #' @param ...    Additional arguments will be ignored.
 #'  
 #' @return An object of class \code{SWIM} containing:
 #'   \itemize{
 #'     \item \code{x}, a data.frame containing the data;
 #'     \item \code{new_weights}, a list, each component corresponds to 
 #'    a different stress and is either a vector of scenario weights or a
 #'    function, that applied to a column of \code{x}, generates the 
 #'    vectors of scenario weights; 
 #'     \item \code{type}, a list, each component corresponds to a 
 #'    different stress and specifies the type of the stress;
 #'     \item \code{specs}, a list, each component corresponds to 
 #'    a different stress and contains a list with the specifications 
 #'    of what has been stressed.
 #'   }
 #' See \code{\link{SWIM}} for details.
 #' 
 #' @author Silvana M. Pesenti 
 #'
 #' @export

  merge.SWIM <- function(x, y, ...){
  if (!is.SWIM(x) | !is.SWIM(y)) stop("x and y are not of class SWIM.")
  if (!identical(get_data(x), get_data(y))) stop("x and y are not based on the same data")
  type <- c(x$type, y$type)
  m <- length(type)
  new_weights <- c(x$new_weights, y$new_weights)
  
  specs <- c(x$specs, y$specs)
  
  # Check if there are duplicate names for stresses
  x_name <- names(x$specs)
  y_name <- names(y$specs)
  
  if (length(intersect(x_name, y_name)) >= 1) {
     names(new_weights) <- paste("stress", 1:m)
     names(specs) <- paste("stress", 1:m)
  } else {
    original <- c(x_name, y_name)
    for (i in 1:length(original)) {
      if (startsWith(original[i], "stress ")) original[i] <- paste("stress", i, collapse = " ")
    }
    names(new_weights) <- original
    names(specs) <- original
  }

  xy <- SWIM("x" = get_data(x), "new_weights" = new_weights, "type" = type, "specs" = specs)
  return(xy)
  }

  #' Merging Two Stressed Models
  #'
  #' This function is a \code{method} for an object of class 
  #'     \code{SWIMw}.
  #'     
  #' @details Merges two objects of class \code{SWIMw}, that 
  #'     are based on the same data. \cr 
  #' 
  #' @param x,y    Objects of class \code{SWIMw}.
  #' @param ...    Additional arguments will be ignored.
  #'  
  #' @return A \code{SWIMw} object containing:
  #'     \itemize{
  #'       \item \code{x}, a data.frame containing the data;
  #'       \item \code{h}, bandwidths;
  #'       \item \code{u}, vector containing the gridspace on [0, 1]
  #'       \item \code{lam}, vector containing the lambda's of the optimized model
  #'       \item \code{str_fY}, function defining the densities of the stressed component;
  #'       \item \code{str_FY}, function defining the distribution of the stressed component;
  #'       \item \code{str_FY_inv}, function defining the quantiles of the stressed component;
  #'       \item \code{gamma}, function defining the risk measure;
  #'       \item \code{new_weights}, a list of functions, that applied to
  #'   the \code{k}th column of \code{x}, generates the vectors of scenario
  #'   weights. Each component corresponds to a different stress;
  #'      \item \code{type}, specifies the stress type
  #'      \item \code{specs}, a list, each component corresponds to
  #'    a different stress and contains a list with the specifications
  #'   of what has been stressed.
  #'     }
  #'     See \code{\link{SWIM}} for details.
  #' @author Zhuomin Mao
  #'
  #' @export  
  merge.SWIMw <- function(x, y, ...){
    if (!is.SWIMw(x) | !is.SWIMw(y)) stop("x and y are not of class SWIMw.")
    if (!identical(get_data(x), get_data(y))) stop("x and y are not based on the same data")
    if (!identical(get_data(x), get_data(y))) stop("x and y are not based on the same gridspace")
    
    type <- c(x$type, y$type)
    h <- c(x$h, y$h)
    lam <- c(x$lam, y$lam)
    str_fY <- c(x$str_fY, y$str_fY)
    str_FY <- c(x$str_FY, y$str_FY)
    str_FY_inv <- c(x$str_FY_inv, y$str_FY_inv)
    gamma <- c(x$gamma, y$gamma)
    m <- length(type)
    new_weights <- c(x$new_weights, y$new_weights)
    specs <- c(x$specs, y$specs)
    # names(new_weights) <- paste("stress", 1:m)
    # names(specs) <- paste("stress", 1:m)
    
    # Check if there are duplicate names for stresses
    x_name <- names(x$specs)
    y_name <- names(y$specs)
    
    if (length(intersect(x_name, y_name)) >= 1) {
      names(new_weights) <- paste("stress", 1:m)
      names(specs) <- paste("stress", 1:m)
      }
     
    xy <- SWIMw("x" = get_data(x), "new_weights" = new_weights, "type" = type, "specs" = specs,
                "str_fY" = str_fY, "str_FY" = str_FY, "str_FY_inv" = str_FY_inv,
                "u" = x$u, "h" = h, "lam"=lam, "gamma"=gamma)
    return(xy)
  }
  
  #' @describeIn get_data extracting summary statistics of scenario weights.
  #'
  #' @return \code{summary_weights}: print a list containing summary statistics 
  #'         of the stresses with each element being a table for a different stress. 
  #'         The summary statistics include minimum, maximum, standard deviation, 
  #'         Gini coefficient, entropy and effective sample size. 
  #'         
  #'         Gini coefficient uses the formula \eqn{\frac{\sum_{i=1}^{n} \sum_{j=1}^{n}\left|x_{i}-x_{j}\right|}{2 n^{2} \bar{x}}}.
  #'         
  #'         Effective Sample Size is equal to n / (1+Var(W)), see 
  #'         Equation (9.13) in Owen, Art B. "Monte Carlo theory, methods and examples." (2013).
  #' @export

  summary_weights <- function(object, wCol = "all"){
    table <- list()
    if (is.character(wCol) && wCol == "all") wCol <- 1:ncol(get_weights(object))
    new_weights <- get_weights(object)[ ,wCol]
    
    for (i in wCol){
      w <- get_weights(object)[, i]
      freqs <- w / length(w)
      sub_table <- t(matrix(c(min(w), max(w), stats::sd(w), .gini(w), .entropy(freqs), length(w)/(1 + stats::var(w)))))
      sub_table <- round(sub_table, 4)
      colnames(sub_table) <- c("min", "max", "sd", "Gini coef", "Entropy", "effective sample size")
      name <- names(object$specs)[i]
      table[[name]] <- sub_table
    }
    print(table)
  }
  
  .gini <- function(w) mean(outer(X = w, Y = w, FUN = function(x, y)abs(x - y))) / 2
  .entropy <- function(freqs) -sum(freqs * log2(freqs))

  
  #' Convert SWIMw to SWIM 
  #'     
  #' @details Convert a SWIMw object into a SWIM object
  #' 
  #' @param object A \code{SWIMw} object
  #'  
  #' @return A \code{SWIM} object containing:
  #'     \itemize{
  #'       \item \code{x}, a data.frame containing the data;
  #'       \item \code{new_weights}, a list of functions, that applied to
  #'   the \code{k}th column of \code{x}, generates the vectors of scenario
  #'   weights. Each component corresponds to a different stress;
  #'      \item \code{type = "VaR"};
  #'      \item \code{specs}, a list, each component corresponds to
  #'    a different stress and contains \code{k},
  #'     \code{alpha} and \code{q}.
  #'     }
  #'     See \code{\link{SWIM}} for details.
  #' 
  #' @examples
  #' \dontrun{
  #' set.seed(0)
  #' x <- as.data.frame(cbind(
  #'   "normal" = rnorm(1000),
  #'   "gamma" = rgamma(1000, shape = 2)))
  #' res1 <- stress_wass(type = "RM", x = x,
  #'   alpha = 0.9, q_ratio = 1.05)
  #' convert_SWIMw_to_SWIM(res1)
  #' }
  #' 
  #' @author Kent Wu 
  #'
  #' @export

  convert_SWIMw_to_SWIM <- function(object) {
    if (!is.SWIMw(object)) stop("Object not of class SWIMw")
    my_list <- SWIM("x" = object$x, "new_weights" = object$x, "type" = object$type, "specs" = object$specs)
    return (my_list)
  }